sys/sched_round_robin/sched_round_robin.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2021 TUBA Freiberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       Round Robin Scheduler implementation
+ *
+ * @author      Karl Fessel <karl.fessel@ovgu.de>
+ *
+ * @}
+ */
+
+#include "sched.h"
+#include "thread.h"
+#include "xtimer.h"
+#include "sched_round_robin.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static void _sched_round_robin_cb(void *d);
+
+static xtimer_t _rr_timer = { .callback = _sched_round_robin_cb };
+
+/*
+ * assuming simple reads from and writes to a byte to be atomic on every board
+ */
+static uint8_t _current_rr_priority = 0;
+
+void sched_runq_callback(uint8_t prio);
+
+void _sched_round_robin_cb(void *d)
+{
+    (void)d;
+    /*
+     * reorder current Round Robin priority
+     * (put the current thread at the end of the run queue of its priority)
+     * and setup the scheduler to schedule when returning from the IRQ
+     */
+    uint8_t prio = _current_rr_priority;
+    if (prio != 0xff) {
+        DEBUG_PUTS("Round_Robin");
+        sched_round_robin_runqueue(prio);
+        _current_rr_priority = 0xff;
+    }
+    thread_t *active_thread = thread_get_active();
+    if (active_thread) {
+        uint8_t active_priority = active_thread->priority;
+        if (active_priority == prio) {
+            thread_yield_higher();
+            /* thread change will call the runqueue_change_cb */
+        }
+        else {
+            sched_runq_callback(active_priority);
+        }
+    }
+}
+
+static inline void sched_round_robin_remove(void)
+{
+    _current_rr_priority = 0xff;
+    xtimer_remove(&_rr_timer);
+}
+
+static inline void sched_round_robin_set(uint8_t prio)
+{
+    _current_rr_priority = prio;
+    xtimer_set(&_rr_timer, SCHED_RR_TIMEOUT);
+}
+
+void sched_runq_callback(uint8_t prio)
+{
+    if (SCHED_RR_MASK & 1 << prio) {
+        return;
+    }
+
+    if (_current_rr_priority == prio) {
+        if (sched_runqueue_len2(prio) == 0) {
+            sched_round_robin_remove();
+            thread_t *active_thread = thread_get_active();
+            if (active_thread) {
+                prio = active_thread->priority;
+            }
+            else {
+                return;
+            }
```

Comment:
```suggestion
            return;
```

Reply:
no, the line line above changes the prio that this funktion is working on and hence (if there is an active prio) will be processed in the second part of that function.

Reply:
?

Reply:
`prio` cannot be `0xff`, right? End because `_current_rr_priority == prio`, we know that `_current_rr_priority` is also not `0xff`. Hence, `_current_rr_priority == 0xff` will evaluate to false in the if condition below.